### PR TITLE
[Doc] How to use Kubectl

### PR DESCRIPTION
**Pull request by eaudetcobello**

### CODE REVIEW

**Comment by evilnick** [comment]
Sorry for snooping, just some comments as you go along:

- please wordwrap the text at 79 chars wherever possible, it makes future diffs and changes much much easier to parse
- I wouldn't include adding an extra node in these docs - is it essential for any of the other things you plan to add? I think it potentially adds some complexity we could do without. We do have a card for 'adding nodes' docs and this would be useful content there though

**Comment by eaudetcobello** [comment]
@evilnick any thoughts so far ?

File: docs/src/tutorial/kubectl.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+## What you will need
+
+Before you begin, make sure you have the following:
+
+- A bootstrapped Canonical K8s cluster (See [insert link]...)
+
+### 1. The Kubectl Command
+
+This commands interacts with the Kubernetes API server (kube-apiserver) and is the most commonly used command when working with Kubernetes, so let's take some time to familiarize ourselves with it.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:11]: I would shorten this sentence to something like this, use an "active voice" and include a refernce:
Use the `kubectl` command to interact with your cluster, establishing a connection with the [Kubernetes API server](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/).

File: docs/src/tutorial/kubectl.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+The `kubectl` command included with Canonical K8s is built from the original upstream source into the `k8s` snap you have installed.
+
+### 2. How To Use Kubectl
+
+You can access kubectl with the following command:
+
+```
+sudo k8s kubectl
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:20]: I would change this to `sudo k8s kubectl <command> ` or point the user to the help section `sudo k8s kubectl --help`

File: docs/src/tutorial/kubectl.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+### 1. The Kubectl Command
+
+This commands interacts with the Kubernetes API server (kube-apiserver) and is the most commonly used command when working with Kubernetes, so let's take some time to familiarize ourselves with it.
+
+The `kubectl` command included with Canonical K8s is built from the original upstream source into the `k8s` snap you have installed.
+
+### 2. How To Use Kubectl
+
+You can access kubectl with the following command:
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:17]: Change this to "active voice">> To access kubectl, run the following command:

File: docs/src/tutorial/kubectl.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+```
+sudo k8s kubectl
+```
+
+Note: Only control plane nodes can use the `kubectl` command. Worker nodes do not have access to this command.
+
+### 3. Configuration
+
+In Canonical K8s, the `kubeconfig` file that is being read to display the configuration when you run `kubectl config view` lives at `/snap/k8s/current/k8s/config/kubeconfig`. You can change this by setting a `KUBECONFIG` environment variable or passing the `--kubeconfig` flag to a command.
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply docs/src/tutorial/kubectl.md:27]: Little side note: For formatting split this section and other long ones over multiple lines. I usually use the "half my screen rule" but don't use it if you're using an ultra-wide monitor ;)

[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply docs/src/tutorial/kubectl.md:27]: @evilnick Would it be appropriate to add a `what gets configured in the kubeconfig` here? 

Either way, I would add a link to the k8s docs kubeconfig: https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/

File: docs/src/tutorial/kubectl.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+### 4. Viewing objects
+
+Let's review what was created in the Getting Started guide.
+
+Let's see what pods were created when we enabled the `network` and `dns` components.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:33]: Alternate sentence beginnings, for example use "To see what pods...".

Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:33]: Let's put enabled `network` and `dns` as a requirement for now until we enable it by default.

File: docs/src/tutorial/kubectl.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+```
+sudo k8s kubectl get pods -o wide -n kube-system
+```
+
+You should be seeing Cilium pods.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:39]: @bschimke95 suggested staying away from using the name of our networking solution. For example, I said: we observe the `network-operator` and `networking-agent` pods running.

File: docs/src/tutorial/kubectl.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+In Canonical K8s, the `kubeconfig` file that is being read to display the configuration when you run `kubectl config view` lives at `/snap/k8s/current/k8s/config/kubeconfig`. You can change this by setting a `KUBECONFIG` environment variable or passing the `--kubeconfig` flag to a command.
+
+### 4. Viewing objects
+
+Let's review what was created in the Getting Started guide.
+
+Let's see what pods were created when we enabled the `network` and `dns` components.
+
+```
+sudo k8s kubectl get pods -o wide -n kube-system
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:36]: We could add a troubleshoot note here: if you see an error message it is likely that you forgot to bootstrap your cluster.

File: docs/src/tutorial/kubectl.md (hunk 6 above)
Comment (louiseschmidtgen) [reply docs/src/tutorial/kubectl.md:39]: and we can see the coredns pod.

File: docs/src/tutorial/kubectl.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+You should be seeing Cilium pods.
+
+```
+sudo k8s kubectl get svc -A
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:42]: Use the unabbreviated command: `sudo k8s kubectl get service --all-namespaces`

File: docs/src/tutorial/kubectl.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+The `kubernetes` service in the `default` namespace is where the Kubernetes API server resides, and it's the endpoint with which other nodes in your cluster will communicate.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:45]: Let's avoid same sentence beginning:
We could structure it as such:
You will observe three services running:
- **coredns**: Provides DNS resolution services.
- **kubernetes**: ...
- **hubble-peer**: ...

File: docs/src/tutorial/kubectl.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+The `hubble-peer` service in the `kube-system` namespace is created by Canonical K8s (an opinionated K8s distribution) to ...
+
+### 5. Creating and Managing Objects
+
+Let's create a deployment using this command:
+
+```
+sudo k8s kubectl create deployment nginx --image=nginx:latest
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:54]: Let's explain the command. For example:
This command launches a [pod](https://kubernetes.io/docs/concepts/workloads/pods/),
the smallest deployable unit in Kubernetes,
running the nginx application within a container.

File: docs/src/tutorial/kubectl.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+Notice how `sudo k8s kubectl get pods` shows you one NGINX pod.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:57]: I would use an "Active voice" here as well and put all commands in a code block:
To observe the NGINX pod running in the default namespace:
```
sudo k8s kubectl get pods
```

File: docs/src/tutorial/kubectl.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+Let's now scale this deployment, which means increasing the number of pods it manages.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:59]: We can sneak in an explanation of a "deployment" and add a link:
To increase the number of NGINX pods, scale the [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) which oversees the pod's management.

File: docs/src/tutorial/kubectl.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+```
+sudo k8s kubectl scale deployment nginx --replicas=3
+```
+
+Run `sudo k8s kubectl get pods` again and notice that you have 3 NGINX pods.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:65]: You can alternate the use of the word "run" with execute for example.

File: docs/src/tutorial/kubectl.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+sudo k8s kubectl get svc -A
+```
+
+The `kubernetes` service in the `default` namespace is where the Kubernetes API server resides, and it's the endpoint with which other nodes in your cluster will communicate.
+
+The `hubble-peer` service in the `kube-system` namespace is created by Canonical K8s (an opinionated K8s distribution) to ...
+
+### 5. Creating and Managing Objects
+
+Let's create a deployment using this command:
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:51]: add a quick note what we are deploying: an nginx server

File: docs/src/tutorial/kubectl.md [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+```
+sudo k8s kubectl scale deployment nginx --replicas=3
+```
+
+Run `sudo k8s kubectl get pods` again and notice that you have 3 NGINX pods.
+
+Let's delete those 3 pods to demonstrate a deployment's ability to ensure the declared state of the cluster is maintained.
+
+Run `sudo k8s kubectl delete pods -l app=nginx`
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:69]: Move this to a separate code block and explain the command: something like: This command removes all pods in the cluster that are labeled with "app=nginx".

File: docs/src/tutorial/kubectl.md [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+If you open another terminal while the above command is executing, you'll notice the original 3 pods will have a status of `Terminating` and 3 new pods will have a status of `ContainerCreating`.
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:71]: We could tell the user to open a window before deleting the pods and to observe them using `sudo k8s kubectl get pods --all-namespace --watch`

File: docs/src/tutorial/kubectl.md [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+If you wish to remove the snap without saving a snapshot of its data, add `--purge` to the command:
+
+```
+sudo snap remove k8s --purge
+```
+This option ensures complete removal of the snap and its associated data.
+
+## Next Steps
+
+- Keep mastering Canonical Kubernetes with kubectl: [How to use kubectl](#TODO)
```
Comment (louiseschmidtgen) [inline docs/src/tutorial/kubectl.md:92]: Remove `How to use kubectl` from the list.

**Review by louiseschmidtgen — CHANGES_REQUESTED** [review]
Great work @eaudetcobello!
I've made a few suggestions about formatting and wording. I walked through your steps without any problems and enjoyed the organization of your steps.

File: docs/src/tutorial/kubectl.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (evilnick) [reply docs/src/tutorial/kubectl.md]: I think this is possibly something to bring up in the daily as it affects UX.

File: docs/src/tutorial/kubectl.md (hunk 10 above)
Comment (eaudetcobello) [reply docs/src/tutorial/kubectl.md:66]: I think how to's can assume familiarity

File: docs/src/tutorial/kubectl.md [hunk 18]
```diff
@@ -0,0 +1,135 @@
+# Basic operations with Kubernetes using kubectl
+
+## What you will need
+
+Before you begin, make sure you have the following:
+
+- A bootstrapped Canonical Kubernetes cluster
+(See [Getting Started](https://github.com/canonical/k8s-snap/blob/main/docs/src/tutorial/getting-started.md))
+
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:9]: I think we should state the assumption that the user doesn't already have another version of  kubectl installed
- You are using the built-in `kubectl` command from the snap

File: docs/src/tutorial/kubectl.md [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+### 2. How To Use Kubectl
+
+To access `kubectl`, run the following command:
+
+```
+sudo k8s kubectl <command>
+```
+
+Note: Only control plane nodes can use the `kubectl` command. Worker nodes do
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:25]: There is a specific style for this, but for the moment the generic MD styling will work for an admonition
```suggestion
> **Note**: Only control plane nodes can use the `kubectl` command. Worker nodes do 
```

File: docs/src/tutorial/kubectl.md [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+In Canonical Kubernetes, the `kubeconfig` file that is being read to display 
+the configuration when you run `kubectl config view` lives at `/snap/k8scurrent/k8s/config/kubeconfig`. You can change this by setting a `KUBECONFIG` 
+environment variable or passing the `--kubeconfig` flag to a command.
+
+To find out more, you can visit 
+[the official kubeconfig documentation](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/)
+
+### 4. Viewing objects
+
+Let's review what was created in the Getting Started guide.
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:39]: add link

File: docs/src/tutorial/kubectl.md [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+To see what pods were created when we enabled the `network` and `dns` 
+components:
+
+```
+sudo k8s kubectl get pods -o wide -n kube-system
+```
+
+You should be seeing the `network-operator`, `networking-agent` and `coredns` pods.
+
+**Note**: If you see an error message here, it is likely that you forgot to
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:50]:
```suggestion
> **Note**: If you see an error message here, it is likely that you forgot to 
```

File: docs/src/tutorial/kubectl.md (hunk 10 above)
Comment (evilnick) [reply docs/src/tutorial/kubectl.md:66]: yes, this isn't a tutorial so it doesn't need to explain everything

File: docs/src/tutorial/kubectl.md [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+```
+sudo k8s kubectl delete pods -l app=nginx
+```
+
+The above command deletes all pods in the cluster that are labeled with `app=nginx`.
+
+You'll notice the original 3 pods will have a status of `Terminating` and 3 
+new pods will have a status of `ContainerCreating`.
+
+### 6. Remove Canonical Kubernetes (Optional)
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:106]: I wouldn't include this part, it's only really valid for the Install  how-tos and the tutorial

File: docs/src/tutorial/kubectl.md [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+
+If you wish to remove the snap without saving a snapshot of its data, add 
+`--purge` to the command:
+
+```
+sudo snap remove k8s --purge
+```
+This option ensures complete removal of the snap and its associated data.
+
+## Next Steps
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:125]:
```suggestion
## Further information
```

File: docs/src/tutorial/kubectl.md [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+
+- Explore Kubernetes commands with our [Command Reference Guide](#TODO)
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:127]: I would remove all these references for now., and just replace with a link to the upstream kubernetes docs and another entry for the k8s snaps commands

**Review by evilnick — CHANGES_REQUESTED** [review]
Thanks @eaudetcobello (and @louiseschmidtgen)
I realise starting out on these docs tasks without a lot of examples to work from or any contributing guide was quite an ask but you have both done great work. 
I have left a few comments to look at, I hope they make sense. I'll take a final look over when they are dealt with :1st_place_medal:

File: docs/src/tutorial/kubectl.md [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+The above command deletes all pods in the cluster that are labeled with
+`app=nginx`.
+
+You'll notice the original 3 pods will have a status of `Terminating` and 3 new
+pods will have a status of `ContainerCreating`.
+
+## Further information
+
+- Explore Kubernetes commands with our [Command Reference Guide](#TODO)
+- See the official kubectl reference
```
Comment (evilnick) [inline docs/src/tutorial/kubectl.md:115]:
```suggestion
- See the official `kubectl` reference
```

**Review by evilnick — APPROVED** [review]
I just added a brief intro - it is illegal and punishable by heavy fines to have two headings with no text in between

**Pull request merged**